File: app/src/main/java/in/eoninfotech/eontechnician/activity/ReceiveDeviceActivity.java
package in.eoninfotech.eontechnician.activity;

import androidx.appcompat.app.AppCompatActivity;
import in.eoninfotech.eontechnician.NonScrollListView;
import in.eoninfotech.eontechnician.R;
import in.eoninfotech.eontechnician.Responses.ClientDetails;
import in.eoninfotech.eontechnician.Responses.ClientLocationResponse;
import in.eoninfotech.eontechnician.Responses.ClientResponse;
import in.eoninfotech.eontechnician.Responses.CollectedItemsResponse;
import in.eoninfotech.eontechnician.Responses.DisconnectionResponse;
import in.eoninfotech.eontechnician.Responses.FaultList;
import in.eoninfotech.eontechnician.Responses.FaultResponse;
import in.eoninfotech.eontechnician.Responses.MainResponse;
import in.eoninfotech.eontechnician.Responses.NotAvailActivityResponse;
import in.eoninfotech.eontechnician.Responses.PaymentMethodResponse;
import in.eoninfotech.eontechnician.Responses.RemovalActivityResponse;
import in.eoninfotech.eontechnician.Responses.RemovalResponse;
import in.eoninfotech.eontechnician.Responses.ReplaceReason;
import in.eoninfotech.eontechnician.Responses.SimOperatorResponse;
import in.eoninfotech.eontechnician.Responses.SimReplaceResponse;
import in.eoninfotech.eontechnician.Responses.VTSResponse;
import in.eoninfotech.eontechnician.Responses.VehNotAvailReasonResponse;
import in.eoninfotech.eontechnician.Responses.VehicleTypeResponse;
import in.eoninfotech.eontechnician.Responses.WorkTypeResponse;
import in.eoninfotech.eontechnician.callbacks.ClientListener;
import in.eoninfotech.eontechnician.controllers.NewInstallmentController;
import in.eoninfotech.eontechnician.view.MySearchableSpinner;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.ListView;

import com.robinhood.ticker.TickerView;

import java.util.ArrayList;

public class ReceiveDeviceActivity extends AppCompatActivity implements ClientListener {

    private TickerView total_device;
    MySearchableSpinner spn_transit, spn_items;
    NonScrollListView device_detail_list;
    EditText other_material, remarks;
    Button btn_accept, btn_reject;
    SharedPreferences sharedprefs;
    SharedPreferences.Editor editor;
    String version, username;
    NewInstallmentController newInstallmentController;
    ArrayList<ClientDetails> clientList = new ArrayList<>();
    ArrayList<String> clientDetail = new ArrayList<>();
    ArrayAdapter<String> adapter;
    ArrayList<FaultList> list_change_values = new ArrayList<>();
    ArrayList<String> value_name = new ArrayList<>();
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_receive_device);
        sharedprefs = getSharedPreferences("login_user_pass", MODE_PRIVATE);
        editor = sharedprefs.edit();
        username = sharedprefs.getString("s_uuser", "");
        version = sharedprefs.getString("version", "");

        initView();
    }
    private void initView() {

        total_device = findViewById(R.id.total_device);
        spn_transit = findViewById(R.id.spn_transit);
        spn_items= findViewById(R.id.spn_items);
        device_detail_list = findViewById(R.id.device_detail_list);
        other_material = findViewById(R.id.other_material);
        btn_accept = findViewById(R.id.btn_accept);
        btn_reject = findViewById(R.id.btn_reject);
        remarks= findViewById(R.id.remarks);
        newInstallmentController = new NewInstallmentController();

        addclients();
        addLocation();
        getFaultList();
        total_device.setPreferredScrollingDirection(TickerView.ScrollingDirection.DOWN);

        btn_accept.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

            }
        });

        btn_reject.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

            }
        });
    }

    private void getFaultList() {
        newInstallmentController.reqeuestFaultList(this);
    }

    private void addLocation() {

    }

    private void addclients() {
        newInstallmentController.reqeuestClientList(this);
    }
    @Override
    public void clientResponse(ClientResponse response) {
        try{
            clientList = response.getClientList();
            try {
                try {
                    clientDetail.clear();
                } catch (Exception e) {
                    e.printStackTrace();
                }   clientDetail.add(" SELECT CLIENT");
                for (int i = 0; i < clientList.size(); i++) {
                    clientDetail.add(clientList.get(i).getClient_Name());
                }
                adapter = new ArrayAdapter<String>(this, R.layout.simple_custom_spinner_item, clientDetail);
                adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
                spn_transit.setAdapter(adapter);
            } catch (NullPointerException npe) {
                npe.printStackTrace();
            }
        }catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public void locationResponse(ClientLocationResponse response) {

    }

    @Override
    public void workTypeResponse(WorkTypeResponse response) {

    }

    @Override
    public void vehicleTypeResponse(VehicleTypeResponse response) {

    }

    @Override
    public void faultListResponse(FaultResponse response) {
        try {
            list_change_values = response.getFaultLists();
            try {
                try {
                    value_name.clear();
                } catch (Exception e) {
                    e.printStackTrace();
                }
                if (list_change_values.size() > 0) {
                    for (int i = 0; i < list_change_values.size(); i++) {
                        value_name.add(list_change_values.get(i).getName());
                    }
                    if (list_change_values.size() > 5) {
                        device_detail_list.setLayoutParams(new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, 80 * list_change_values.size() + 1));
                    } else {
                        device_detail_list.setLayoutParams(new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, 80 * list_change_values.size()));
                    }
                    adapter = new ArrayAdapter<String>(this, R.layout.simple_custom_list_item, value_name);
                    device_detail_list.setChoiceMode(ListView.CHOICE_MODE_MULTIPLE);
                    device_detail_list.setAdapter(adapter);
                } else {
                }
            } catch (NullPointerException npe) {
                npe.printStackTrace();
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public void replaceResponse(ReplaceReason response) {

    }

    @Override
    public void disconnectionResponse(DisconnectionResponse response) {

    }

    @Override
    public void removalActivityResponse(RemovalActivityResponse response) {

    }

    @Override
    public void removalResponse(RemovalResponse response) {

    }

    @Override
    public void damageResponse(RemovalResponse response) {

    }

    @Override
    public void collectItemResponse(CollectedItemsResponse response) {

    }

    @Override
    public void simOperatorResponse(SimOperatorResponse response) {

    }

    @Override
    public void simReplaceReason(SimReplaceResponse response) {

    }

    @Override
    public void notAvailActivity(NotAvailActivityResponse response) {

    }

    @Override
    public void vehicleNotAvailReason(VehNotAvailReasonResponse response) {

    }

    @Override
    public void vtsResponses(VTSResponse response) {

    }

    @Override
    public void vtsResponse(VTSResponse response) {

    }

    @Override
    public void pMethod(PaymentMethodResponse response) {

    }

    @Override
    public void updateDataResponse(MainResponse response) {

    }
}

File: app/src/main/java/in/eoninfotech/eontechnician/activity/ReceiveDeviceActivity.kt
package `in`.eoninfotech.eontechnician.activity

import android.app.AlertDialog
import android.app.DatePickerDialog
import android.content.SharedPreferences
import android.graphics.Color
import android.os.Bundle
import android.text.InputType
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import dmax.dialog.SpotsDialog
import `in`.eoninfotech.eontechnician.R
import `in`.eoninfotech.eontechnician.ReceiveDeviceAdapter
import `in`.eoninfotech.eontechnician.responses.DispatchDeviceDetails
import `in`.eoninfotech.eontechnician.responses.DispatchDeviceList
import `in`.eoninfotech.eontechnician.responses.MainResponse
import `in`.eoninfotech.eontechnician.callbacks.ReceiveDeviceListener
import `in`.eoninfotech.eontechnician.controllers.ReceiveDeviceController
import `in`.eoninfotech.eontechnician.databinding.ActivityReceiveDeviceBinding
import `in`.eoninfotech.eontechnician.helper.CheckConnection
import `in`.eoninfotech.eontechnician.helper.K
import `in`.eoninfotech.eontechnician.view.MySearchableSpinner
import java.text.SimpleDateFormat
import java.util.*


class ReceiveDeviceActivity : AppCompatActivity(), ReceiveDeviceListener {

    private var receiveDeviceAdapter: ReceiveDeviceAdapter? = null
    private var binding: ActivityReceiveDeviceBinding? = null
    private var sharedprefs: SharedPreferences? = null
    private var editor: SharedPreferences.Editor? = null
    private var version: String? = null
    private var username: String? = null
    private var techid: String? = null
    private var status: String? = "A"
    private var current_date: String? = null
    private var s_from_date: String? = null
    private var s_to_date: String? = null
    private lateinit var fromdate: EditText
    private lateinit var todate: EditText
    private lateinit var spinner: MySearchableSpinner
    private var chk = CheckConnection(this)
    private var calen = Calendar.getInstance()
    private var receiveDeviceController: ReceiveDeviceController? = null
    private var progressDialog: AlertDialog? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityReceiveDeviceBinding.inflate(layoutInflater)
        setContentView(binding!!.root)

        sharedprefs = getSharedPreferences("login_user_pass", MODE_PRIVATE)
        editor = sharedprefs!!.edit()
        username = sharedprefs!!.getString("s_uuser", "")
        version = sharedprefs!!.getString("version", "")
        techid = sharedprefs!!.getString("s_user_id", "")

        supportActionBar?.apply {
            setDisplayHomeAsUpEnabled(true)
            title = "Receive Devices"
        }

        setDate()
        initView()
    }

    override fun onSupportNavigateUp(): Boolean {
        onBackPressedDispatcher.onBackPressed()
        return true
    }

    private fun setDate() {
        val year = calen.get(Calendar.YEAR)
        var month = calen.get(Calendar.MONTH) + 1
        val day = calen.get(Calendar.DAY_OF_MONTH)
        current_date = "$day-${if (month < 10) "0$month" else "$month"}-$year"

        binding!!.fromDate.setText(current_date)
        binding!!.toDate.setText(current_date)

        s_from_date = K.getDateFormatWithMonthNameHyphenYear(current_date)
        s_to_date = K.getDateFormatWithMonthNameHyphenYear(current_date)
    }

    private fun initView() {
        receiveDeviceAdapter = ReceiveDeviceAdapter(this, arrayListOf())

        binding!!.recyclerView.apply {
            layoutManager = LinearLayoutManager(this@ReceiveDeviceActivity)
            adapter = receiveDeviceAdapter
        }

        binding!!.fromDate.inputType = InputType.TYPE_NULL
        binding!!.toDate.inputType = InputType.TYPE_NULL

        receiveDeviceController = ReceiveDeviceController()
        progressDialog = SpotsDialog(this, R.style.CustomInstallation)

        binding!!.swipeRefresh.setOnRefreshListener {
            getReceiveData()
        }

        setupDatePickers()
        setupStatusSpinner()
    }

    private fun setupDatePickers() {
        val c = Calendar.getInstance()
        val year = c.get(Calendar.YEAR)
        val month = c.get(Calendar.MONTH)
        val day = c.get(Calendar.DAY_OF_MONTH)

        binding!!.fromDate.setOnClickListener {
            val datePickerDialog = DatePickerDialog(this, { _, y, m, d ->
                val selectedDate = "$d-${m + 1}-$y"
                binding!!.fromDate.setText(selectedDate)
                s_from_date = K.getDateFormatWithMonthNameHyphenYear(selectedDate)
                getReceiveData()
            }, year, month, day)

            datePickerDialog.datePicker.maxDate = System.currentTimeMillis()
            datePickerDialog.show()
        }

        binding!!.toDate.setOnClickListener {
            val datePickerDialog = DatePickerDialog(this, { _, y, m, d ->
                val selectedDate = "$d-${m + 1}-$y"
                binding!!.toDate.setText(selectedDate)
                s_to_date = K.getDateFormatWithMonthNameHyphenYear(selectedDate)
                getReceiveData()
            }, year, month, day)

            datePickerDialog.datePicker.maxDate = System.currentTimeMillis()
            datePickerDialog.show()
        }
    }

    private fun setupStatusSpinner() {
        binding!!.statusSpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>, view: View?, position: Int, id: Long) {
                val text = parent.getItemAtPosition(position).toString()
                status = when (text.lowercase()) {
                    "received" -> "R"
                    "not received" -> "NR"
                    else -> "A"
                }
                getReceiveData()
            }

            override fun onNothingSelected(parent: AdapterView<*>) {}
        }
    }

    private fun getReceiveData() {
        if (chk.isConnected) {
            progressDialog?.show()
            receiveDeviceController?.requestReceiveDevice(s_from_date, s_to_date, status, techid, this)
        } else {
            chk.showConnectionErrorDialog()
        }
    }

    override fun receiveDeviceResponse(response: MainResponse) {
        progressDialog?.dismiss()
        binding!!.swipeRefresh.isRefreshing = false

        val newList = response.dispatched_device_list ?: arrayListOf()

        if (response.type == 1 && newList.isNotEmpty()) {
            receiveDeviceAdapter?.updateList(newList)
            binding!!.recyclerView.visibility = View.VISIBLE
            binding!!.txtContentUnavailable.visibility = View.GONE
        } else {
            binding!!.recyclerView.visibility = View.GONE
            binding!!.txtContentUnavailable.visibility = View.VISIBLE
        }
    }

    override fun receiveDispatchMaterial(response: MainResponse?) {}
    override fun returnDeviceresponse(response: MainResponse?) {}
    override fun dispatchFromTechResponse(response: MainResponse?) {}

    override fun onResume() {
        super.onResume()
        getReceiveData()
    }

}